### PR TITLE
Push Defined Functions List into DeclChecker

### PR DESCRIPTION
**Pull request by CodaFi**
Pull from the previous attempt to do this, but take a less radical tact and make the list local to the DeclChecker because there are still ordering issues with witness matching.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

File: lib/Sema/TypeCheckDeclPrimary.cpp [hunk 1]
```diff
@@ -1166,8 +1166,18 @@ namespace {
...
+  std::vector<AbstractFunctionDecl *> definedFunctions;
 
   explicit DeclChecker(ASTContext &ctx) : Ctx(ctx) {}
+  ~DeclChecker() {
+    for (AbstractFunctionDecl *AFD : definedFunctions) {
+      TypeChecker::typeCheckAbstractFunctionBody(AFD);
+    }
+
+    for (AbstractFunctionDecl *FD : llvm::reverse(definedFunctions)) {
+      TypeChecker::computeCaptures(FD);
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDeclPrimary.cpp]: It should be fine to combine these two loops into one. We're no longer adding nested local functions to definedFunctions.

Comment (CodaFi) [reply lib/Sema/TypeCheckDeclPrimary.cpp]: This patch doesn't work as-is. I'm going to force-push at this PR and use it to kill TypeChecker instead.

**Pull request closed without merging**